### PR TITLE
[WIP] refactor to prepare digitization

### PR DESCRIPTION
**Pull request by massich**
The idea of this PR was to move all the code that seems to be dealing with digitization in a single place before start changing code and factoring similar functions.

But I get in dependency hell.

Any help is wellcome. @agramfort @larsoner

### CODE REVIEW

**Comment by massich** [comment]
Maybe I should propose a different refractor where io stays in io. And at some point we can get rid of having our structures inside io

**Comment by larsoner** [comment]
This pull request **introduces 11 alerts** when merging 32f17c8bb5d0cd71b8f9aafa5945145c11ea5ca7 into 3d12b2b7b3b0ddb3f50962e4293cddf8ccf6ba52 - [view on LGTM.com](https://lgtm.com/projects/g/mne-tools/mne-python/rev/pr-72f3cfdbc838df623afd5c5d9d46835b63256e14)

**new alerts:**

* 6 for Use of the return value of a procedure
* 3 for Unused import
* 1 for Wrong name for an argument in a call
* 1 for Wrong number of arguments in a call

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by larsoner** [comment]
That would probably be easier. Moving things out of io makes problems harder

**Comment by agramfort** [comment]
I agree with @larsoner I would keep digitization.py in mne/io to avoid circular import mess

**Comment by larsoner** [comment]
This pull request **introduces 4 alerts** when merging 6c4423ee7fe5245db19698ff48c913f7c1e22957 into ce50c2b20ac1197d79e68665900c9f6646505b13 - [view on LGTM.com](https://lgtm.com/projects/g/mne-tools/mne-python/rev/pr-bb55fba70b9ba580b49210308e097498f76b0840)

**new alerts:**

* 4 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by larsoner** [comment]
This pull request **introduces 1 alert** when merging b8218bb49dc3f39b25e3acf4151aaa3687d67008 into 4555011717b9008f41eaa4926d598bb570e42fca - [view on LGTM.com](https://lgtm.com/projects/g/mne-tools/mne-python/rev/pr-19cf5f31a9eb63b0e2c08cabbffcc20eed8334b9)

**new alerts:**

* 1 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by larsoner** [comment]
This pull request **introduces 3 alerts** when merging 50e46f09e0b70410a9936c0094e9b7615a46f5b6 into b3dfd7a1dd4cfc2fb501f9db9c4bcb3eb158cc2c - [view on LGTM.com](https://lgtm.com/projects/g/mne-tools/mne-python/rev/pr-a91c63eb5dd8abc44e96caeef6e0b7b0711b1649)

**new alerts:**

* 1 for Special method has incorrect signature
* 1 for Unused import
* 1 for Unreachable code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by massich** [comment]
At d1eaa8d, `Digitization` is a custom list of `DigPoint` all boilerplate.

@larsoner do you think that its worth to factor it out? I would like to be able to write something like (if I understand how Bunch works):

```
from ..utils import Bunch
from ..utils import MyListOfBunch

class DigPoint(Bunch):

    def __repr__(self):  # noqa: D105
        # maybe we need to write this one

    def __eq__(self, other):
        # this one should come for free


class Digitization(MyListOfBunch, Digpoint):
    """Represent a list of DigPoint objects.

    Parameters
    ----------
    elements : list
        A list of DigPoint objects.
    """
    pass # because everything comes for free
```

**Comment by larsoner** [comment]
This pull request **introduces 3 alerts** when merging 9fe020b655206da18d77b73b877faaca68a4025d into b3dfd7a1dd4cfc2fb501f9db9c4bcb3eb158cc2c - [view on LGTM.com](https://lgtm.com/projects/g/mne-tools/mne-python/rev/pr-65471367486731f0943b166019642a37e1ecfb36)

**new alerts:**

* 1 for Special method has incorrect signature
* 1 for Unused import
* 1 for Unreachable code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by katarinaslama** [comment]
The travis error seems to not be specific to this PR. (When I run make pep on the current mne dev branch, I get the same error.) Is there anything I could help with here?

**Comment by larsoner** [comment]
I don't get errors with `make pep` on `master`. The tracebacks I see on travis have to do with `mne/digitization` which is new. @katarinaslama can you paste the `master` error here (or open a PR to fix it)?

It's possible one of the PEP checkers updated and got more strict

**Comment by katarinaslama** [comment]
> I don't get errors with `make pep` on `master`. The tracebacks I see on travis have to do with `mne/digitization` which is new. @katarinaslama can you paste the `master` error here (or open a PR to fix it)?
> 
> It's possible one of the PEP checkers updated and got more strict

It's also possible that I'm missing something. My read of the [travis output](https://travis-ci.org/mne-tools/mne-python/jobs/522215622) suggests that lines 3386-3391 are the problem.

When I run `make pep` on `master`, I get the same output as line 3390:
`make: *** [pep] Error 2`

Am I misreading this? Thanks!

File: mne/digitization/base.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
+# -*- coding: utf-8 -*-
+# Authors: Alexandre Gramfort <alexandre.gramfort@telecom-paristech.fr>
+#          Eric Larson <larson.eric.d@gmail.com>
+#          Joan Massich <mailsik@gmail.com>
+#
+# License: BSD (3-clause)
+from collections.abc import MutableMapping, MutableSequence
+from mne.transforms import _coord_frame_name
+from mne.io.constants import FIFF
+from copy import deepcopy
```
Comment (agramfort) [inline mne/digitization/base.py]: wrong order of imports
also you should use relative imports

**Comment by larsoner** [comment]
The failure it's much earlier in line 3324. make flake (which is called by make pep) fails on this PR, but it doesn't get reported until all pep targets are run, so you have to look father up to see it

**Comment by larsoner** [comment]
This pull request **introduces 2 alerts** when merging ce3c3ef3ba0f15919a2b30cf4ab4d4892d303bb8 into 9cc5d1457c29151eea73fa7f14619724182960c1 - [view on LGTM.com](https://lgtm.com/projects/g/mne-tools/mne-python/rev/pr-3c860add1341b8fa41f234c4af84bdef0325eb12)

**new alerts:**

* 1 for Special method has incorrect signature
* 1 for Unreachable code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by larsoner** [comment]
This pull request **introduces 2 alerts** when merging 890a23244981799c61355d2950087389cab4aa9f into 92f61faef72a335d6de63e816bd5bcfeaceffa09 - [view on LGTM.com](https://lgtm.com/projects/g/mne-tools/mne-python/rev/pr-021e4b8d5514a94abcc3261d00658297ddc612da)

**new alerts:**

* 1 for Special method has incorrect signature
* 1 for Unreachable code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by larsoner** [comment]
This pull request **introduces 2 alerts** when merging ea07437902cac0faf6063b4409e5cdfe2cd99016 into 92f61faef72a335d6de63e816bd5bcfeaceffa09 - [view on LGTM.com](https://lgtm.com/projects/g/mne-tools/mne-python/rev/pr-989d94675eae17e069c5a3be39a465a752fa32aa)

**new alerts:**

* 1 for Special method has incorrect signature
* 1 for Unreachable code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by massich** [comment]
* Progress:
  - [ ] io/artemis123/artemis123.py
  - [ ] io/brainvision/brainvision.py
  - [ ] io/bti/bti.py
  - [ ] io/cnt/cnt.py
  - [ ] io/ctf/ctf.py
  - [ ] io/edf/edf.py
  - [ ] io/edf/edf.py
  - [ ] io/edf/edf.py
  - [ ] io/eeglab/eeglab.py
  - [ ] io/egi/egi.py
  - [ ] io/eximia/eximia.py
  - [ ] io/fieldtrip/fieldtrip.py
  - [ ] io/fiff/raw.py
  - [x] io/kit/kit.py
  - [ ] io/nicolet/nicolet.py
  - [ ] realtime/client.py
 
* Things that we found and we need to come back to (maybe new PRs):
  - General:
    - Epochs constructors have none homogeneous signature.
    
  - io/kit/kit.py:
    - Epochs and Raw __init__ should be almost the same, and they are not.

**Comment by larsoner** [comment]
Excellent, thanks for figuring this out

**Comment by larsoner** [comment]
This pull request **introduces 2 alerts** when merging b5c55044f844163960d6c43dc1129b23d192bf18 into 2dd7aa0bd05f3cae8dc5c17e951454ef6c71b17c - [view on LGTM.com](https://lgtm.com/projects/g/mne-tools/mne-python/rev/pr-7586079bac0964de3cf5288d6917bc1f4f8b28d6)

**new alerts:**

* 1 for Special method has incorrect signature
* 1 for Unreachable code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by larsoner** [comment]
This pull request **introduces 2 alerts** when merging b65cc2236b6d48c68bf607a9343a057b7a173902 into 11d91e7cd2ea6264da355b47dd8fc8b200bc10db - [view on LGTM.com](https://lgtm.com/projects/g/mne-tools/mne-python/rev/pr-155f53636c03ada1d8f78290677ae86034b37182)

**new alerts:**

* 1 for Special method has incorrect signature
* 1 for Unreachable code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by larsoner** [comment]
This pull request **introduces 2 alerts** when merging 4cc18b14b35f89f61841e4b77cfeb8aadd949d40 into 5a9ae3c3d66b6275fffec217cf560acf651913d7 - [view on LGTM.com](https://lgtm.com/projects/g/mne-tools/mne-python/rev/pr-6e126b814fe4fc0258973178c1c8d4d6d2e83a21)

**new alerts:**

* 1 for Special method has incorrect signature
* 1 for Unreachable code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by larsoner** [comment]
This pull request **introduces 3 alerts** when merging f2684c1efe05ba82bbbe928b6fe6b42d514138d0 into 5a9ae3c3d66b6275fffec217cf560acf651913d7 - [view on LGTM.com](https://lgtm.com/projects/g/mne-tools/mne-python/rev/pr-f48d7dcd97a587afbde5e3b0a8d00b0d6d9a68f5)

**new alerts:**

* 1 for Special method has incorrect signature
* 1 for Unused import
* 1 for Unreachable code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by larsoner** [comment]
This pull request **introduces 2 alerts** when merging e003129034d056fe270e65cf716d7f90f3dbc2ee into b94eff339a5be230fce74ead7e562f7813b8c40d - [view on LGTM.com](https://lgtm.com/projects/g/mne-tools/mne-python/rev/pr-c7561cfb70ea9da13e411352d8567e74e53031e2)

**new alerts:**

* 1 for Special method has incorrect signature
* 1 for Unreachable code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by larsoner** [comment]
This pull request **introduces 2 alerts** when merging 3fcf2b5c3809256ef2ba11af1dbda1c52c93b5ba into c4a83c5a1792fa68315baa87f1d7b26d9b6d709b - [view on LGTM.com](https://lgtm.com/projects/g/mne-tools/mne-python/rev/pr-fa2b70cfd4a01eaad8cdd1dd1ef9c44ee2281946)

**new alerts:**

* 1 for Special method has incorrect signature
* 1 for Unreachable code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by larsoner** [comment]
This pull request **introduces 2 alerts** when merging 577900ce8a635d151d9eafc857052b9782343a3e into b9dcbe1eac621e22b5cff6883e4b3a889846e303 - [view on LGTM.com](https://lgtm.com/projects/g/mne-tools/mne-python/rev/pr-74f09a7531690f81508af96b966ee37873f0b185)

**new alerts:**

* 1 for Special method has incorrect signature
* 1 for Unreachable code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by larsoner** [comment]
This pull request **introduces 2 alerts** when merging 3fcf2b5c3809256ef2ba11af1dbda1c52c93b5ba into 03bc75e9067483166bb8ebbb2827ccaf45425031 - [view on LGTM.com](https://lgtm.com/projects/g/mne-tools/mne-python/rev/pr-ab0d53bfc7d9cceac5f858fa306ce8a0931da9cc)

**new alerts:**

* 1 for Special method has incorrect signature
* 1 for Unreachable code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by massich** [comment]
```py
    _msg = f'{raw.__class__.__name__} needs Digitization'
    assert isinstance(raw.info['dig'], (type(None), Digitization)), _msg
    ```
    
    - [ ] AssertionError: RawArtemis123 needs Digitization
    - [ ] AssertionError: RawBrainVision needs Digitization
    - [ ] AssertionError: RawBTi needs Digitization
    - [ ] AssertionError: RawCTF needs Digitization
    - [ ] AssertionError: RawEDF needs Digitization
    - [ ] AssertionError: RawEEGLAB needs Digitization
    - [ ] AssertionError: Raw needs Digitization
    - [ ] AssertionError: RawKIT needs Digitization

**Pull request closed without merging**